File: src/main/java/com/booknara/practice/kotlin/annotation/JvmOverload.java
package com.booknara.practice.kotlin.annotation;

import java.util.Date;

public class JvmOverload {
    public JvmOverload() {
        JvmOverloadKt.bar();    // Compile error without @JvmOverloads
        JvmOverloadKt.bar(1, 0.1, "Custom string");

        Event event1 = new Event();
        Event event2 = new Event("Birthday party"); // Compile error without @JvmOverloads
        Event event3 = new Event("Celebration", new Date(),true);
    }
}

File: src/main/java/com/booknara/practice/kotlin/annotation/JvmOverload.kt
//@file:JvmName("SomeName") // to change the file class name
package com.booknara.practice.kotlin.annotation

import java.util.*

fun main() {
    bar()
    bar(1)
    bar(1, 0.1)
    bar(1, 0.1, "Custom string")

    Event("Birthday party")
    Event("Celebration", Date(), true)
}

@JvmOverloads
fun bar(a:Int = 0, b: Double = 0.0, c:String = "default value") {
    println("a=$a, b=$b, c = $c")
}

data class Event @JvmOverloads constructor(var name: String? = "",
                                           var date: Date? = Date(),
                                           var private: Boolean = false)